src/sentry/utils/services.py

```diff
@@ -250,7 +262,23 @@ def execute(*args, **kwargs):
                     '{!r} is not a registered backend.'.format(
                         selected_backend_names[0]))
 
-            call_backend_method = operator.methodcaller(attribute_name, *args, **kwargs)
+            def call_backend_method(backend):
+                base = self.__backend_base
+                active_backends = type(self).state.backends
+
+                # Ensure that we haven't somehow accidentally entered a context
+                # where the backend we're calling has already been marked as
+                # active (or worse, some other backend is already active.)
+                assert base not in active_backends
```

Comment:
I guess this will *prevent* mix/matching, right?

I can't think of a reason we'd want it, but it's possible some other usecase would. We can change it then, though.

EDIT: Errr, by "this" I mean this asserts it and the code above prevents it.

Reply:
> I guess this will *prevent* mix/matching, right?

Can you be more specific here about what you mean by mix/matching? I don't think I understand.

Reply:
Oh right, this is keyed by base, disregard me, I think. I was reading this like we *couldn't* have a Snuba search hit a non-Snuba tagstore internally (or vice versa). But this just prevents a Snuba search from doing a non-Snuba search internally.

Reply:
Yep, yep. This basically just checks to make sure that a thread that is supposed to be bound to a specific backend doesn't unexpectedly change. I considered writing a "this should never happen" comment next to this assertion but I know better than to tempt fate like that.